zmq-lwt/src/zmq_lwt.ml

```diff
@@ -0,0 +1,82 @@
+let (>>=) = Lwt.(>>=)
+
+module Socket = struct
+
+  type 'a t = {
+    socket : 'a ZMQ.Socket.t;
+    fd : Lwt_unix.file_descr;
+  }
+
+  exception Break_event_loop
+
+  let of_socket socket = {
+    socket;
+    fd = Lwt_unix.of_unix_file_descr ~blocking:false ~set_flags:false (ZMQ.Socket.get_fd socket);
+  }
+
+  let to_socket s = s.socket
+
+  (* Wrap possible exceptions and events which can occur in a ZeroMQ call *)
+  let wrap f s =
+    let io_loop () =
+      Lwt_unix.wrap_syscall Lwt_unix.Read s.fd (
+        fun () ->
+          (* Check for zeromq events *)
+          match ZMQ.Socket.events s.socket with
+          | ZMQ.Socket.No_event -> raise Lwt_unix.Retry
+          | ZMQ.Socket.Poll_in
```

Comment:
In the recv case we should only wait for Poll_in / Poll_in_out. If zmq announces that we can now write to the socket, the read will return nothing. I think that is not what we want. The same goes for send.